File: problems/Maximum Width of Binary Tree/widthOfBinaryTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int WidthOfBinaryTree(TreeNode root) {
        var bfs = (new [] {new { Node = root, Depth = 0, Position = 0 }}).ToList();
        var depth = 0;
        var left = 0;
        var result = 0;

        while (0 < bfs.Count) {
            var curr = bfs[0];
            bfs.RemoveAt(0);

            if (null != curr.Node) {
                bfs.Add(new { Node = curr.Node.left, Depth = 1 + curr.Depth, Position = curr.Position << 1 });
                bfs.Add(new { Node = curr.Node.right, Depth = 1 + curr.Depth, Position = 1 + (curr.Position << 1) });

                if (depth != curr.Depth) {
                    depth = curr.Depth;
                    left = curr.Position;
                }

                result = Math.Max(result, 1 + curr.Position - left);
            }
        }

        return result;
    }
}

File: problems/Maximum Width of Binary Tree/widthOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        return 1 + getMaxWidth(root, 1, new ArrayList<Integer>(), new ArrayList<Integer>(), 1);
    }

    int getMaxWidth(TreeNode root, int level, ArrayList<Integer> start, ArrayList<Integer> end, int pos) {
        if (null == root) {
            return 0;
        }

        if (start.size() < level) {
            start.add(pos);
            end.add(pos);
        } else {
            end.set(level - 1, pos);
        }

        int left = getMaxWidth(root.left, 1 + level, start, end, pos*2-1);
        int right = getMaxWidth(root.right, 1 + level, start, end, pos*2);
        int cur = end.get(level - 1) - start.get(level - 1);

        return Math.max(cur, Math.max(left, right));

    }
}
